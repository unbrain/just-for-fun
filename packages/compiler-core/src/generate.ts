export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  const functionName = `render`
  const args = ['_ctx', '_cache', '$props', '$setup', '$data', '$options']
  const signature = args.join(', ')

  push(`export function `)
  push(`${functionName}(${signature}) {`)
  const node = ast.codegenNode

  push(`return `)
  genNode(ast.codegenNode, context)
  push('}')

  return {
    code: context.code,
  }
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
  }
  return context
}
function genNode(node, content) {
  const { push } = content
  push(`'${node.content}'`)
}
