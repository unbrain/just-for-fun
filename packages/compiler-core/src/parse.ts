import { NodeTypes } from './ast'

interface Context {
  source: string
  // children: any[],
}

enum TagTypes {
  START,
  END,
}

export function baseParse(content: string) {
  const context = createParseContent(content)
  return createRoot(parseChildren(context, []))
}

function parseChildren(context: Context, ancestors: string[]) {
  const nodes: any[] = []
  while (!isEnd(context, ancestors)) {
    const s = context.source
    let node
    if (s.startsWith('{{')) { node = parseInterpolation(context) }
    else if (s.startsWith('<')) {
      if (/[a-z]/i.test(context.source[1]))
        node = parseElement(context, ancestors)
    }
    else {
      node = parseText(context)
    }
    nodes.push(node)
  }
  return nodes
}

function isEnd(context: Context, ancestors) {
  const s = context.source
  for (let i = ancestors.length - 1; i >= 0; i--) {
    const tag = ancestors[i]
    if (startsWithEndTagOpen(s, tag))
      return true
  }
  return !s
}

function parseText(context: Context) {
  let endIndex = context.source.length
  const endToken = ['{{', '</']
  for (let i = 0; i < endToken.length; i++) {
    const index = context.source.indexOf(endToken[i])
    if (index !== -1 && endIndex > index)
      endIndex = index
  }

  const content = parseTextData(context, endIndex)
  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseTextData(context, length) {
  const content = context.source.slice(0, length)

  advanceBy(context, content.length)

  return content
}

function parseElement(context, ancestors: string[] = []) {
  const element: any = parseTag(context, TagTypes.START)
  ancestors.push(element.tag)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()
  if (startsWithEndTagOpen(context.source, element.tag))
    parseTag(context, TagTypes.END)
  else
    throw new Error('tag is not closed')

  return element
}

function startsWithEndTagOpen(source, tag) {
  return source.startsWith('</') && source.slice(2, 2 + tag.length) === tag
}

function parseTag(context, type: TagTypes) {
  const match: any = /^<\/?([a-z]*)>/i.exec(context.source)
  const tag = match[1]
  advanceBy(context, match[0].length)

  if (type === TagTypes.END)
    return
  return {
    type: NodeTypes.ELEMENT,
    tag,
  }
}

function advanceBy(context, length) {
  context.source = context.source.slice(length)
}

function parseInterpolation(context) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter)
  if (closeIndex === -1)
    return
  const rawContentLength = closeIndex - closeDelimiter.length
  advanceBy(context, openDelimiter.length)
  const rawContent = parseTextData(context, rawContentLength)
  const content = rawContent.trim()
  advanceBy(context, closeDelimiter.length)
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  }
}

function createRoot(children) {
  return {
    children,
  }
}

function createParseContent(content: string) {
  return {
    source: content,
  }
}
