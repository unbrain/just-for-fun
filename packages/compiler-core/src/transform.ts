import { NodeTypes } from './ast'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)

  createRootCodegen(root, context)
}

function createRootCodegen(root, context) {
  const child = root.children[0]
  // if (child.type === NodeTypes.ELEMENT)
  root.codegenNode = child
  // else
  // root.codegenNode = createVNodeCall(context, root, child)
}

function traverseNode(node, context) {
  context.nodeTransfoms.forEach((item) => {
    item(node)
  })
  traverseChildren(node, context)
}

function traverseChildren(node, context) {
  const children = node.children
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traverseNode(node, context)
    }
  }
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransfoms: options.nodeTransfoms || [],
  }

  return context
}
