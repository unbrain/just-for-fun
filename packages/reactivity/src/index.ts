export { ref, proxyRefs } from './ref'
export { shallowReadonly, reactive, readonly, isReactive, isReadonly, isProxy } from './reactive'
export { effect } from './effect'
