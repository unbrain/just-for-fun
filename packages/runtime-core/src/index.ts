export { h } from './h'
export { renderSlots } from '../helper/renderSlots'
export { createTextVNode } from './vnode'
export { getCurrentInstance } from './components'
export { provide, inject } from './apiInject'
export { createRenderer } from './render'
