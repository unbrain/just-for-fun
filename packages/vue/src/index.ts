export * from '@zy/runtime-dom'
export * from '@zy/reactivity'
