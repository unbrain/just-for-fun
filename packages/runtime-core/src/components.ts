import { proxyRefs, shallowReadonly } from '@zy/reactivity'
import { initSlots } from './componentSlots'
import { emit } from './componentEmits'
import { initProps } from './componentProps'
import { PublicInstanceProxyHandlers } from './componentPublicInstance'

export function createComponentInstance(vnode, parent) {
  const instance = {
    vnode,
    parent,
    type: vnode.type,
    setupState: {},
    el: null,
    slots: {},
    provides: parent ? parent.provides : {},
    isMounted: false,
    emit: (_string, ..._args) => { },
  }
  
  instance.emit = emit.bind(null, instance)

  return instance
}

export function setupComponent(instance) {
  // TODO: initProps initSlots
  initProps(instance, instance.vnode.props)
  initSlots(instance, instance.vnode.children)
  setupStatefulComponent(instance)
}

function setupStatefulComponent(instance) {
  const { type, props, emit } = instance
  const Component = type
  instance.proxy = new Proxy({ _: instance }, PublicInstanceProxyHandlers)
  const { setup } = Component

  if (setup) {
    setCurrentInstance(instance)
    const setupResult = setup(shallowReadonly(props), { emit })
    setCurrentInstance(null)

    handleSetupResult(instance, setupResult)
  }
}
let currentInstance = null
export function getCurrentInstance() {
  return currentInstance
};

function setCurrentInstance(instance) {
  currentInstance = instance
};

function handleSetupResult(instance: any, setupResult) {
  // TODO: function
  if (typeof setupResult === 'object')
    instance.setupState = proxyRefs(setupResult)

  finishComponentSetup(instance)
}

function finishComponentSetup(instance: any) {
  const Component = instance.type
  // if (Component?.render) {
  instance.render = Component.render
  // }
}
