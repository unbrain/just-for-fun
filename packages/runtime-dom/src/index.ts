import { createRenderer } from '@zy/runtime-core'

export function createElement(type) {
  return document.createElement(type)
}

export function patchProps(el, key, _val, nextVal) {
  const isOn = /on[A-Z]/.test(key)
  if (isOn) {
    const event = key.slice(2).toLocaleLowerCase()
    el.addEventListener(event, nextVal)
  }
  else {
    if (nextVal === null || nextVal === undefined)
      el.removeAttribute(key)
    else
      el.setAttribute(key, nextVal)
  }
}

export function insert(child, container, anchor) {
  container.insertBefore(child, anchor || null)
}

export function remove(el) {
  const parent = el.parentNode
  if (parent)
    parent.removeChild(el)
}

export function setElementText(el, text) {
  el.textContent = text
}

const renderer: any = createRenderer({
  createElement,
  patchProps,
  insert,
  remove,
  setElementText,
})

export function createApp(...args) {
  return renderer.createApp(...args)
}

export * from '@zy/runtime-core'
