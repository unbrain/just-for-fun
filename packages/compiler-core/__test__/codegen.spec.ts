import { describe, expect, it } from 'vitest'
import { baseParse } from '../src/parse'
import { generate } from '../src/generate'
import { transform } from '../src/transform'

describe('codegen', () => {
  it('codegen', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)

    expect(code).toMatchSnapshot()
  })
})
