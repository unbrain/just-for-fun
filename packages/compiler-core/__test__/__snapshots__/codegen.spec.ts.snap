// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > codegen 1`] = `"export function render(_ctx, _cache, $props, $setup, $data, $options) {return 'hi'}"`;
